### DISCUSSION TITLE
how to serialize non featured data type?

### DISCUSSION THREAD

**Discussion by Sombian** [Minestom/Minestom#1870] [Help] [answered]
in `Spigot`'s API, one create custom `PersistentDataType` to store almost anything, easily.

e.g.

```kotlin
abstract class NamedBinaryTag<T : Any, Z : Any> : PersistentDataType<T, Z> {
	abstract fun toPrimitive(complex: Z): T
	
	override fun toPrimitive(complex: Z, context: PersistentDataAdapterContext): T {
		return toPrimitive(complex)
	}
	abstract fun fromPrimitive(primitive: T): Z

	override fun fromPrimitive(primitive: T, context: PersistentDataAdapterContext): Z {
		return fromPrimitive(primitive)
	}
}
class _SERIALIZABLE<T : Serializable>(private val type: Class<T>) : NamedBinaryTag<ByteArray, T>() {
	override fun getPrimitiveType(): Class<ByteArray> {
		return ByteArray::class.java
	}
	override fun getComplexType(): Class<T> {
		return this.type
	}
	override fun toPrimitive(complex: T): ByteArray {
		val bos = ByteArrayOutputStream()
		val oos = ObjectOutputStream(bos)
		oos.writeObject(complex)
		oos.flush()
		return bos.toByteArray()
	}
	override fun fromPrimitive(primitive: ByteArray): T {
		val bis = ByteArrayInputStream(primitive)
		val ois = ObjectInputStream(bis)
		@Suppress("UNCHECKED_CAST")
		return ois.readObject() as T
	}
}
```

certainly, current `Tag` API provides extra types that are not present in `Spigot`'s by default.
but `Minestorm` also dont provides some useful types like `ByteArray` that can be found in `Spigot`'s.

im very new to `Minestom`, hence thats why im asking,
since there arent much tutorials or resources out there yet.

can i have some examples of how to achieve this?

**Comment by TheMode** [comment]
You can use `Tag.NBT("name")` for array support (and probably even cast the generic to be `NBTByteArray`). Or if you are fine with a list, `Tag.Byte("name").list()`

This is only in answer to your byte array question, feel free to ask if you wonder about another specific type

**Reply by Sombian** [reply]
does API support any type out of box?
and how does `Tag.NBT("name")` magically supports array?

```java
    public static @NotNull Tag<NBT> NBT(@NotNull String key) {
        return tag(key, Serializers.NBT_ENTRY);
    }
```

i dont see any hard-coded string from the source code.

```java
public class Cosmetic implements Serializable {
	private final Material material;
	private final Integer model;

	Cosmetic(final @NotNull Material material, final @Nullable Integer model) {
		this.material = material;
		this.model = model;
	}

	public Material getMaterial() {
		return this.material;
	}

	public Integer getModel() {
		return this.model;
	}
}
```

i want to serialize this for example. thanks in advance!

**Reply by TheMode** [reply, marked as answer]
The Tag API is ultimately a way to interact with a NBT Compound, so you can do anything using `Tag.NBT` and there is a dedicated Byte Array type present in the format: https://minecraft.fandom.com/wiki/NBT_format#Data_types

You would have to do something like
```java
Tag<NBTByteArray> tag = Tag.class.cast(Tag.NBT("path"));

var handler = TagHandler.newHandler();
NBTByteArray byteArray = handler.getTag(tag);
```

For enums we do not have an util method for it, you have to choose between serializing the enum name using `Tag.String`, or its ordinal with `Tag.Integer`
```java
// Store name
var handler = TagHandler.newHandler();
var tag = Tag.String("enum");
handler.setTag(tag, Test.A.name());

// Store ordinal
var handler = TagHandler.newHandler();
var tag = Tag.Integer("enum");
handler.setTag(tag, Test.A.ordinal());

// Store ordinal but use enum
var handler = TagHandler.newHandler();
Tag<Test> tag = Tag.Integer("enum").map(v -> Test.values()[v], Enum::ordinal);
handler.setTag(tag, Test.A);
```

As for your `Cosmetic` example I would first recommend to never use the `Serializable` interface, and you can use it in the tag API as follow
```java
public record Cosmetic(String material, int model)  {
}
var handler = TagHandler.newHandler();
Tag<Cosmetic> tag = Tag.Structure("cosmetic", Cosmetic.class);
handler.setTag(tag, new Cosmetic("Something", 5));
Cosmetic cosmetic = handler.getTag(tag);
```

**Reply by Sombian** [reply]
thank you for your kind and fast responeses!
im really glad i can play without stupid NMS in `Minestom`'s favour!
